// mask/lib/mask.js
/**
 * mask.js - a dead simple mask
 */

(function (definition) {
    if (typeof exports === "object") {
        module.exports = definition();
    } else {
        Mask = definition();
    }

})(function () {

    function setDefaultStyle(el, style) {
        for (var attr in style) el.style[attr] = style[attr];
    }

    var on = (function () {
        if("addEventListener" in window) {
            return function (el, evt, fn, bubble) {
                el.addEventListener(evt, fn, bubble);
            };
        }
        else if("attachEvent" in window) {
            return function (el, evt, fn, bubble) {
                el.attachEvent("on" + evt, fn);
            };
        }
    }());

    var off = (function () {
        if("removeEventListener" in window) {
            return function (el, evt, fn) {
                el.removeEventListener(evt, fn);
            };
        }
        else if("detachEvent" in window) {
            return function (el, evt, fn) {
                el.detachEvent("on" + evt, fn);
            };
        }
    }());

    function Zopacityf(visible, d) {
        return Zanimo.f('opacity', visible ? 0.5 : 0, d, 'ease-in-out');
    }

    var Q = window.Q || require('q'),
        Zanimo = window.Zanimo || require('zanimo'),
        inputEvent = window.document.ontouchstart === null ? "touchstart": "click",
        Mask = {},
        defaultStyle = {
            width : '100%',
            height : '100%',
            position : 'absolute',
            top : 0,
            left : 0,
            opacity : 0,
            display : 'none',
            backgroundColor : 'black',
            zIndex : 10
        },
        duration = 500,
        initialized = false,
        displayed = false,
        animated = false,
        transitionf = Zopacityf,
        el = null;

    Mask.init = function (config) {
        if (initialized) return;
        el = window.document.createElement('div');

        if (config) {
            if (config.id) el.id = config.id;
            if (config.backgroundColor)
                defaultStyle.backgroundColor = config.backgroundColor;
            if (config.duration) duration = config.duration;
            if (config.transitionf) transitionf = config.transitionf;
            if (config.zIndex) defaultStyle.zIndex = config.zIndex;
        }

        setDefaultStyle(el, defaultStyle);
        window.document.body.appendChild(el);
        initialized = true;
    };

    Mask.style = function (style) { setDefaultStyle(el, style); };

    Mask.show = function () {
        if(!initialized) return Q.reject("Mask is not initialized!");
        if(animated) return Q.reject("Mask is in use!");
        if(displayed) return Q.reject("Mask is already displayed!");
        animated = true;
        return Zanimo(el)
            .then(Zanimo.f('display', 'block'))
            .then(transitionf(true, duration))
            .then(function (el) {
                on(el, inputEvent, Mask.f);
                return el;
            })
            .fin(function(el) {
                displayed = true;
                animated = false;
                return el;
            });
    };

    Mask.hide = function () {
        if(!initialized) return Q.reject("Mask is not initialized!");
        if(animated) return Q.reject("Mask is in use!");
        if(!displayed) return Q.reject("Mask is already hidden!");
        animated = true;
        return Zanimo(el)
            .then(transitionf(false, duration))
            .then(Zanimo.f('display', 'none'))
            .then(function (el) {
                off(el, inputEvent, Mask.f);
                return el;
            })
            .fin(function(el) {
                displayed = false;
                animated = false;
                return el;
            });
    };

    Mask.onTouch = function(f) { Mask.f = f; };

    return Mask;
});
